stl/inc/variant

```diff
@@ -1616,10 +1584,37 @@ constexpr _Ret _Visit_impl(_Callable&& _Obj, _Variants&&... _Args) {
                             : _Size <= 256 ? 4
                                            : -1;
     return _Visit_strategy<_Strategy>::template _Visit2<_Ret, _ListOfIndexVectors>(
-        _Variant_visit_index1(0, static_cast<_As_variant<_Variants>&>(_Args)...), static_cast<_Callable&&>(_Obj),
+        _STD _Variant_visit_index1(0, static_cast<_As_variant<_Variants>&>(_Args)...), static_cast<_Callable&&>(_Obj),
         static_cast<_As_variant<_Variants>&&>(_Args)...);
 }
 
+template <class _Expected, class _Callable, class _ArgList, class... _Variants>
+inline constexpr bool _Variant_all_visit_results_same = false;
+
+template <class _Expected, class _Callable, class... _Args>
+inline constexpr bool _Variant_all_visit_results_same<_Expected, _Callable, _Meta_list<_Args...>> =
+    is_same_v<decltype(_STD declval<_Callable>()(_STD declval<_Args>()...)), _Expected>;

```

Comment:
This appears to be a regression: `_STD declval<_Callable>()(_STD declval<_Args>()...)` is using the function call operator instead of the `invoke` protocol (as the old machinery did).

Also occurs in `_Variant_all_visit_results_convertible`.

This also indicates a lack of test coverage for PMFs/PMDs.

Reply:
Regression confirmed:

```
D:\GitHub\STL\out\build\x64>type meow.cpp
```
```cpp
#include <iostream>
#include <variant>
using namespace std;

struct Point2D {
    int x;
    int y;
};

struct Point3D : Point2D {
    int z;
};

int main() {
    using Var = variant<Point2D, Point3D>;

    Var v1{Point2D{10, 20}};
    Var v2{Point3D{30, 40, 50}};

    cout << visit(&Point2D::y, v1) << endl;
    cout << visit(&Point2D::y, v2) << endl;
}
```

Works with 17.4 Preview 1:

```
D:\GitHub\STL\out\build\x64>cl /EHsc /nologo /W4 /std:c++latest meow.cpp && meow
meow.cpp
20
40
```

Error with this PR:

```
D:\GitHub\STL\out\build\x64>cl /EHsc /nologo /W4 /std:c++latest meow.cpp && meow
meow.cpp
D:\GitHub\STL\out\build\x64\out\inc\variant(1596): error C2064: term does not evaluate to a function taking 1 arguments
D:\GitHub\STL\out\build\x64\out\inc\variant(1601): note: see reference to variable template 'const bool _Variant_all_visit_results_same<int &,int Point2D::*,std::_Meta_list<Point2D &> >' being compiled
D:\GitHub\STL\out\build\x64\out\inc\variant(1627): note: see reference to variable template 'const bool _Variant_all_visit_results_same<int &,int Point2D::*,std::_Meta_list<>,std::variant<Point2D,Point3D> &>' being compiled
meow.cpp(20): note: see reference to function template instantiation 'int &std::visit<intPoint2D::* ,Var&,void>(_Callable &&,Var &)' being compiled
        with
        [
            _Callable=int Point2D::*
        ]
D:\GitHub\STL\out\build\x64\out\inc\variant(1595): error C2064: term does not evaluate to a function taking 1 arguments
D:\GitHub\STL\out\build\x64\out\inc\variant(1595): error C2062: type 'unknown-type' unexpected
D:\GitHub\STL\out\build\x64\out\inc\variant(1596): error C3376: 'std::_Variant_all_visit_results_same<_Expected,_Callable,std::_Meta_list<_Args...>,>': only static data member templates are allowed
D:\GitHub\STL\out\build\x64\out\inc\variant(1601): error C2131: expression did not evaluate to a constant
D:\GitHub\STL\out\build\x64\out\inc\variant(1601): note: failure was caused by a read of an uninitialized symbol
D:\GitHub\STL\out\build\x64\out\inc\variant(1601): note: see usage of 'std::?$_Variant_all_visit_results_same@$RBAAB@$RBAAC@U?$_Meta_list@$$W$TBAAD@@std@@$$V<_Ret,intPoint2D::* ,std::_Meta_list<Point2D &>,>'
D:\GitHub\STL\out\build\x64\out\inc\variant(1627): error C2131: expression did not evaluate to a constant
D:\GitHub\STL\out\build\x64\out\inc\variant(1601): note: failure was caused by a read of an uninitialized symbol
D:\GitHub\STL\out\build\x64\out\inc\variant(1601): note: see usage of 'std::?$_Variant_all_visit_results_same@$RBAAB@$RBAAC@U?$_Meta_list@$$W$TBAAD@@std@@$$V<_Ret,intPoint2D::* ,std::_Meta_list<Point2D &>,>'
```

Reply:
Just a bit of a GIANT HOLE in our test coverage.